### PR TITLE
ISPN-7696 RemoveExpired could be called with updated value

### PR DESCRIPTION
**Pull request by wburns**
https://issues.jboss.org/browse/ISPN-7696

### CODE REVIEW

File: core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java [hunk 1]
```diff
@@ -76,10 +76,14 @@ public void processExpiration() {
...
                if (e.canExpire()) {
-                  if (ExpiryHelper.isExpiredMortal(e.getLifespan(), e.getCreated(), currentTimeMillis)) {
-                     handleLifespanExpireEntry(e, true);
-                  } else if (ExpiryHelper.isExpiredTransient(e.getMaxIdle(), e.getLastUsed(), currentTimeMillis)) {
-                     super.handleInMemoryExpiration(e, currentTimeMillis);
+                  // Have to synchronize on the entry to make sure we see the value and metadata at the same time
+                  synchronized (e) {
+                     long lifespan = e.getLifespan();
+                     if (ExpiryHelper.isExpiredMortal(lifespan, e.getCreated(), currentTimeMillis)) {
+                        handleLifespanExpireEntry(e.getKey(), e.getValue(), lifespan, true);
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java]: You don't need to call this within the synchronized block.

Comment (wburns) [inline core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java]: I am aware. I was being more lazy and to be honest it doesn't save very much as this lock is only on the entry. I can change it up though.

Comment (wburns) [inline core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java]: And actually I also would need synchronization around the `else if` block, which is probably why I didn't separate it. Code will be come quite messy. I will see if I can clean it up more though.

**Comment by wburns** [comment]
Fixed.

**Comment by rvansa** [comment]
Great, let's integrate this.

**Comment by ryanemerson** [comment]
@wburns Integrated, thanks!

**Pull request merged**